[Tree Borrows] Full Tree Borrows Support (#221)
minirust/minirust: spec/mem/tree_borrows/memory.md

```rust
pub struct TreeBorrowsMemory<T: Target> {
    allocations: List<TreeBorrowsAllocation>,
    /// Map from active call IDs to their protecting nodes.

>     active_calls: Map<CallId, List<TreeBorrowsProvenance>>,

Reviewer: There seems to be a state invariant here: a `TreeBorrowsProvenance` is in the list for some `CallId` if and only if it is currently (weakly or strongly) protected.

That should be documented. Also, I wonder if we can avoid that redundancy...

    _phantom: std::marker::PhantomData<T>,
}

> pub struct TreeBorrowsFrameExtra {

Reviewer: ```suggestion
/// Our per-frame state is the list of nodes that are protected by this call.
pub struct TreeBorrowsFrameExtra {
```

        protected: Protected,
        frame_extra: &mut TreeBorrowsFrameExtra,
    ) -> Result<Pointer<TreeBorrowsProvenance>> {
        let Some((alloc_id, parent_path)) = ptr.provenance else {
            // An invalid pointer.

>             throw_ub!("Tree Borrows: reborrowing pointer without provenance");

Reviewer: I don't think this should be UB... how can it be that the tests still pass?^^

This should just do nothing and return the pointer unchanged. Please also add a test for this case, something that creates a `*mut ()` via `ptr::without_provenance_mut` and then turns that into a reference and passes that to a function.

        frame_extra: &mut TreeBorrowsFrameExtra,
    ) -> Result<Pointer<TreeBorrowsProvenance>> {
        let Some((alloc_id, parent_path)) = ptr.provenance else {

>             if pointee_size.is_zero() { return ret(ptr);}

Reviewer: ```suggestion
            // Pointers without provenance cannot access any memory, so giving them a new
            // tag makes no sense. If the pointee also has size zero, this is fine, otherwise UB.
            if pointee_size.is_zero() { return ret(ptr);}
```

>             throw_ub!("Tree Borrows: reborrowing a pointer without provenance");

Reviewer: ```suggestion
            throw_ub!("Tree Borrows: non-zero-sized reborrow of a pointer without provenance");
```

        frame_extra: &mut TreeBorrowsFrameExtra,
    ) -> Result<Pointer<TreeBorrowsProvenance>> {
        let Some((alloc_id, parent_path)) = ptr.provenance else {

>             return ret(ptr);

Reviewer: ```suggestion
            // If the pointer has no provenance, just keep it unchanged. It can't be used to access any memory anyway.
            return ret(ptr);
```

Reviewer: Hm wait... this is only okay if `pointee_size` is zero. If it is non-zero, this should be UB!

Author: oops, I added a UB test case for this...

https://github.com/minirust/minirust/pull/221/files#diff-14483848f1f65ed137acedd3856e275ef52b1b2516bdd8e84424941890f6c7ba

            ret::<Result<Path>>(child_path)
        })?;

        // Track the new protector
        if protected.yes() { self.add_protector(call_id, (alloc_id, child_path)); }

        // Create the child pointer and return it 
        ret(Pointer {
            provenance: Some((alloc_id, child_path)),
            ..ptr
        })
    }

    /// Remove the protector.
    /// `provenance` is the provenance of the protector.
    /// Perform a special implicit access on all locations that have been accessed.
    fn release_protector(&mut self, provenance: TreeBorrowsProvenance) -> Result {
        let (alloc_id, path) = provenance;
        self.allocations.mutate_at(alloc_id.0, |allocation| {

>             // Special case for the `Box`, since the `Box` could be deallocated during the execution of the function

Reviewer: ```suggestion
            // If this is weakly protected, it could have been already deallocated.
            // In that case, we don't do anything here -- fake reads from freed memory don't
            // make a lot of sense.
```

            ret::<Result<Path>>(child_path)
        })?;

        // Track the new protector
        if protected.yes() { self.add_protector(call_id, (alloc_id, child_path)); }

        // Create the child pointer and return it 
        ret(Pointer {
            provenance: Some((alloc_id, child_path)),
            ..ptr
        })
    }

    /// Remove the protector.
    /// `provenance` is the provenance of the protector.
    /// Perform a special implicit access on all locations that have been accessed.
    fn release_protector(&mut self, provenance: TreeBorrowsProvenance) -> Result {
        let (alloc_id, path) = provenance;
        self.allocations.mutate_at(alloc_id.0, |allocation| {
            let protected_node = allocation.extra.root.get_node(path)?;
            // If the protector is weak (i.e. from a `Box` argument) and the location was deallocated
            // there is no need to peform the implicit here.

>             if protected_node.protected == Protected::Weak && !allocation.live { return ret(()); }

Reviewer 2: shoud we have this panic if the allocation is dead but the protector is strong?

    /// Return the provenance of the pointer and offset of the pointer in the allocation.
    fn check_ptr(&self, ptr: Pointer<TreeBorrowsProvenance>, len: Size) -> Result<Option<(TreeBorrowsProvenance, Size)>> {
        // We do not even have to check for null, since no allocation will ever contain that address.
        // Now try to access the allocation information.
        let Some((alloc_id, path)) = ptr.provenance else {
            // We allow zero-sized accesses for `None` provenance.
            // (Provenance monotonicity says that if we allow this,
            // we have to allow it for all provenance.)

>             if len.is_zero() { return ret(None); }

Reviewer: I understand now why you changed this, but it's not quite the right behavior. Probably retagging should not use check_ptr.

Please add two tests:
- it should be UB to violate the protector even with a zero-sized pointee, if it has provenance
- it should *not* be UB to have a zero-sized pointer that is outside the bounds of the allocation

        allocation.deallocation_check(ptr.addr, kind, size, align)?;

        // Check that ptr has the permission to write the entire allocation.
        allocation.extra.root.access(Some(path), AccessKind::Write, Offset::ZERO, size)?;

        // Check that allocation is not strongly protected.

>         if allocation.extra.root.contain_strong_protector() {

Reviewer: Alternatively to this, we could have an `AccessKind::Dealloc` and have `access` itself also check for strong protectors. That would avoid a second traversal over the full tree. I am honestly not sure what is better... @sun-jacobi  @JoJoDeveloping any opinion?

Author: This makes sense to me :)

Reviewer: Okay, let's keep this as-is then. :)

>         if allocation.extra.root.contains_strong_protector() {

Reviewer: So this will make it UB to deallocate memory even if there is a zero-sized strongly protected reference, right? Does that match Miri / Coq?
Cc @JoJoDeveloping

Reviewer 2: It matches Coq, idk about miri (but @sun-jacobi can test that himself).

Author: It also matches Miri :)

Reviewer: In Miri this does not seem to be considered UB, at least under Stacked Borrows:
```rust
use std::alloc::{alloc, dealloc, Layout};

fn test(_x: &mut (), ptr: *mut u8, l: Layout) {
    unsafe { dealloc(ptr, l) };
}

fn main() {
    let l = Layout::from_size_align(1, 1).unwrap();
    let ptr = unsafe { alloc(l) };
    unsafe {test(&mut *ptr.cast::<()>(), ptr, l);}
}
```

Reviewer: But it *is* UB under Tree Borrows...

Reviewer: Okay so let's match Miri TB for now, though I think we may want to change this in the future.

        allocation.deallocation_check(ptr.addr, kind, size, align)?;

        // Check that ptr has the permission to write the entire allocation.
        allocation.extra.root.access(Some(path), AccessKind::Write, Offset::ZERO, size)?;

>         // Check that allocation is not strongly protected.

Reviewer: ```suggestion
        // Check that allocation is not strongly protected.
        // TODO: This makes it UB to deallocate memory even if the strong protector covers 0 bytes!
        // That's different from SB, and we might want to change it in the future.
```

        // Check that ptr has the permission to write the entire allocation.

>         allocation.extra.root.access(Some(path), AccessKind::Write, Size::ZERO, size)?;

Reviewer 2: ```suggestion
        allocation.extra.root.access(Some(path), AccessKind::Write, Offset::ZERO, size)?;
```

`Offset` is a type alias for `Size`, using it here increases readability.

```rust
impl<T: Target> Memory for TreeBorrowsMemory<T> {
    fn retag_ptr(
        &mut self,
        ptr: Pointer<Self::Provenance>,
        ptr_type: PtrType,
        fn_entry: bool,
        frame_extra: &mut Self::FrameExtra,
    ) -> Result<Pointer<Self::Provenance>> {
        match ptr_type {

>             PtrType::Ref { mutbl, pointee } if !pointee.freeze && mutbl == Mutability::Immutable => ret(ptr),

Reviewer: ```suggestion
            PtrType::Ref { mutbl, pointee } if !pointee.freeze && mutbl == Mutability::Immutable => {
                // Shared reference to interior mutable type: retagging is a NOP.
                ret(ptr)
            }
```
